### ISSUE TITLE
Clarify transaction synchronizer semantics [SWF-54]

### ISSUE THREAD

**Issue by spring-operator** [spring-projects/spring-webflow#988]
**[Keith Donald](https://jira.spring.io/secure/ViewProfile.jspa?name=kdonald)** opened **[SWF-54](https://jira.spring.io/browse/SWF-54?redirect=false)** and commented

Use of the "TransactionSynchronizer" strategy is not completely clear and arguably doesn't need to be exposed to user code.

Specifically, since every "signalEvent" request into a FlowExecution is synchronized ("synchronize on conversation"), there is no thread safety issue -- a request signaling an event in a flow execution will be processed in full before another request is allowed to process.

Furthermore, if the second request is logically a duplicate request of a previous one (e.g. from hitting the submit button really fast, or going back and pressing submit), the flow may not allow it for several reasons:
- the flow execution may have transitioned  to another view state and that state may not support the same event (NoMatchingTransitionException)
- The flow may have ended (NoSuchFlowExecution exception).

If the flow transitioned to another view state that does support the event you will end up signaling the event in the WRONG state which will lead to strange results.

The Itemlist sample demonstrates a solution to this by having the client send in a explicit "_stateId" that tells SWF what state to signal the event in.  So there is no possibility of signaling an event in the 'wrong' view state when going 'back" without telling the server.  Specifically, even if one request into SWF results in a state transition, explicitly providing the 'stateId' will cause SWF to skip to the expected state before signaling the 'duplicate' event.  Itemlist also sends along with a "_transactionId" to account for the duplicate submit.

The sellItem sample takes a different approach and demonstrates transactions with continuations, where the flow itself cares for asserting whether it has logically 'completed' yet or not.

Basically, we need to take a careful look at the transaction synch. facility -- look to build it into the framework as much as possible and provide consistent usage semantics that don't expose its workings to user code.

---
No further details from [SWF-54](https://jira.spring.io/browse/SWF-54?redirect=false)
